boa/src/builtins/array/mod.rs

```diff
@@ -957,6 +957,77 @@ impl Array {
         Ok(Value::from(false))
     }
 
+    /// `Array.prototype.reduce( callbackFn [ , initialValue ] )`
+    ///
+    /// The reduce method traverses left to right starting from the first defined value in the array,
+    /// accumulating a value using a given callback function. It returns the accumulated value.
+    ///  
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///  - [MDN documentation][mdn]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-array.prototype.reduce
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce
+    pub(crate) fn reduce(
+        this: &Value,
+        args: &[Value],
+        interpreter: &mut Interpreter,
+    ) -> ResultValue {
+        if args.is_empty() {
+            return Err(Value::from(
+                "missing callback when calling function Array.prototype.reduce",
+            ));
+        }
+        let callback = args.get(0).cloned().unwrap_or_else(Value::undefined);
```

Comment:
```suggestion
		let callback = match args.get(0) {
			Some(value) if value.is_funtion() => value,
			_ => return interperter.throw_type_error("missing callback when calling function Array.prototype.reduce"),
		};
```

```rust
Err(Value::from("missing callback when calling function Array.prototype.reduce"));
```
This throws a string not a `TypeError`, to throw a we use the `Interpreter::throw_type_error(msg)` also if callback is undefined the a `TypeError` should be thrown